app/demo-geant-integration/PrimaryGeneratorAction.cc

```diff
@@ -7,21 +7,54 @@
 //---------------------------------------------------------------------------//
 #include "PrimaryGeneratorAction.hh"
 
-#include <G4ParticleTable.hh>
-#include <G4SystemOfUnits.hh>
-#include <G4ThreeVector.hh>
+#include "corecel/Macros.hh"
+#include "accel/ExceptionConverter.hh"
+#include "accel/HepMC3PrimaryGenerator.hh"
 
-#include "HepMC3Reader.hh"
+#include "GlobalSetup.hh"
+
+using celeritas::HepMC3PrimaryGenerator;
 
 namespace demo_geant
 {
+namespace
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Global HepMC3 file reader shared across threads.
+ *
+ * The first time this is called, the reader will be initialized from the
+ * GlobalSetup event file argument.
+ */
+HepMC3PrimaryGenerator& shared_reader()
+{
+    static celeritas::HepMC3PrimaryGenerator reader{
+        GlobalSetup::Instance()->GetEventFile()};
+    return reader;
+}
+//---------------------------------------------------------------------------//
+} // namespace
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the total number of events available in the HepMC3 file.
+ *
+ * This will load the HepMC3 file if not already active.
+ */
+int PrimaryGeneratorAction::NumEvents()
+{
+    return shared_reader().NumEvents();
+}
```

Comment:
This might be a silly question, but why do we need a duplicate function for the number of events, if we could just call `shared_reader().NumEvents()`? Or simply construct a `static HepMC3PrimaryGenerator` in `demo-geant-integration.cc::run(...)`

Reply:
It's simply an easy way to hide the implementation... none of the other classes in `demo-geant-integration` need to access any of the reader data except for the primary generator (to emit primaries) and the main function (to count the number of events). This is imho the best way of encapsulating it.

You can't put  the primary generator in `demo-geant-integration.cc`  because it needs to be accessed by this class.

Reply:
(also `shared_reader` is just a local anonymous function inside the .cc so no one else can see it.)

Reply:
Yeah that's fair. And having only the concrete implementation `G4VUserPrimaryGeneratorAction` being visible in `main` and using it during `BeamOn` makes more sense than before (which called `HepMC3Reader::Instance()` to fetch the number of events).